controlplane/kubeadm/internal/etcd/etcd.go

```diff
@@ -0,0 +1,233 @@
+}
+
+// MemberAlarm represents an alarm type association with a cluster member.
+type MemberAlarm struct {
+	// MemberID is the ID of the member associated with the raised alarm.
+	MemberID uint64
+
+	// Type is the type of alarm which has been raised.
+	Type AlarmType
+}
+
+type AlarmType int32
+
+const (
+	// AlarmOK denotes that the cluster member is OK.
+	AlarmOk AlarmType = iota
+
+	// AlarmNoSpace denotes that the cluster member has run out of disk space.
+	AlarmNoSpace
+
+	// AlarmCorrupt denotes that the cluster member has corrupted data.
+	AlarmCorrupt
+)
+
+// Adapted from kubeadm
+
+// Member struct defines an etcd member; it is used to avoid spreading
+// github.com/coreos/etcd dependencies.
+type Member struct {
+	// ID is the ID of this cluster member
+	ID uint64
+
+	// Name is the human-readable name of the member. If the member is not started, the name will be an empty string.
+	Name string
+
+	// PeerURLs is the list of URLs the member exposes to the cluster for communication.
+	PeerURLs []string
+
+	// ClientURLs is the list of URLs the member exposes to clients for communication. If the member is not started, clientURLs will be empty.
+	ClientURLs []string
+
+	// IsLearner indicates if the member is raft learner.
+	IsLearner bool
+
+	// Alarms is the list of alarms for a member.
+	Alarms []AlarmType
+}
+
+// pbMemberToMember converts the protobuf representation of a cluster member to a Member struct.
+func pbMemberToMember(m *etcdserverpb.Member) *Member {
+	return &Member{
+		ID:         m.GetID(),
+		Name:       m.GetName(),
+		PeerURLs:   m.GetPeerURLs(),
+		ClientURLs: m.GetClientURLs(),
+		IsLearner:  m.GetIsLearner(),
+		Alarms:     []AlarmType{},
+	}
+}
+
+// NewEtcdClient creates a new etcd client with a custom dialer and is configuration with optional functions.
+func NewEtcdClient(endpoint string, dialer GRPCDial, tlsConfig *tls.Config) (*clientv3.Client, error) {
+	etcdClient, err := clientv3.New(clientv3.Config{
+		Endpoints:   []string{endpoint},
+		DialTimeout: etcdTimeout,
+		DialOptions: []grpc.DialOption{
+			grpc.WithBlock(), // block until the underlying connection is up
+			grpc.WithContextDialer(dialer),
+		},
+		TLS: tlsConfig,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to create etcd client")
+	}
+	etcdClient.Endpoints()
+	return etcdClient, nil
+}
+
+// NewClientWithEtcd configures our response formatter (Client) with an etcd client and endpoint.
+func NewClientWithEtcd(etcdClient etcd) (*Client, error) {
+	if len(etcdClient.Endpoints()) == 0 {
+		return nil, errors.New("etcd client was not configured with any endpoints")
+	}
+	return &Client{
+		Endpoint:   etcdClient.Endpoints()[0],
+		EtcdClient: etcdClient,
+	}, nil
+}
+
+// Close closes the etcd client.
+func (c *Client) Close() error {
+	return c.EtcdClient.Close()
+}
+
+// Member retrieves the Member information for a given peer.
+func (c *Client) Member(ctx context.Context, peerURL string) (*Member, error) {
```

Comment:
Where do we plan to use this?

I can imagine returning the etcd member data based on the etcd member _name_ would be useful, because if you know the Machine, you can find the NodeRef, whose Name will be equal to its etcd member name.

Reply:
Should this be defined in the [`etcd` interface](https://github.com/kubernetes-sigs/cluster-api/pull/2237/files#diff-2563ae029ccd4516e81325b980a0d822R33-R43)?

Reply:
What if  we make this a filter function, e.g., `MemberForName(members []*Member, name string) *Member`?

Reply:
Can you expand on that a little more? What's the use case?

Reply:
Regarding your first comment -- I am more than happy to delete code that isn't useful! If you don't need this function we should absolutely delete it. I was keeping the interface almost the same as the original PR and looking for exactly this feedback.

If you don't need it, we should remove it.